### PR TITLE
Introduce manifest readers and writers

### PR DESCRIPTION
**Pull request by tsreaper**

### CODE REVIEW

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileReader.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    private final Configuration conf;
+    private final ManifestEntrySerializer serializer;
+    private final BulkFormat<RowData, FileSourceSplit> readerFactory;
+
+    public ManifestFileReader(
+            Configuration conf,
+            RowType partitionType,
+            RowType rowType,
+            RowType keyType,
+            Function<RowType, BulkFormat<RowData, FileSourceSplit>> readerFactoryCreator) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileReader.java]: Can we create a class `BulkFormatFactory`?
`Function<RowType, BulkFormat<RowData, FileSourceSplit>>` is too long...

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileWriter.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    private final RowType partitionType;
+    private final ManifestEntrySerializer serializer;
+    private final BulkWriter.Factory<RowData> writerFactory;
+
+    public ManifestFileWriter(
+            RowType partitionType,
+            RowType rowType,
+            RowType keyType,
+            Function<RowType, BulkWriter.Factory<RowData>> writerFactoryCreator) {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileWriter.java]: Ditto, create a factory?

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileReader.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.io.IOException;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * Reader for a manifest file. This file includes several {@link ManifestEntry}s, representing the
+ * additional changes since last snapshot.
+ */
+public class ManifestFileReader {
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileReader.java]: I am thinking can we have a `ManifestFile` class:
- method1: write
- method2: read

They seem to be together all the time.

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileWriter.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    public ManifestFileMeta write(List<ManifestEntry> entries, Path path) throws IOException {
+        Preconditions.checkArgument(
+                entries.size() > 0, "Manifest entries to write must not be empty.");
+
+        FSDataOutputStream out =
+                path.getFileSystem().create(path, FileSystem.WriteMode.NO_OVERWRITE);
+        BulkWriter<RowData> writer = writerFactory.create(out);
+        long numAddedFiles = 0;
+        long numDeletedFiles = 0;
+        FieldStatsCollector collector = new FieldStatsCollector(partitionType);
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestFileWriter.java]: statsCollector

File: flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestListWriter.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+    public ManifestListWriter(
+            RowType partitionType,
+            Function<RowType, BulkWriter.Factory<RowData>> writerFactoryCreator) {
+        this.serializer = new ManifestFileMetaSerializer(partitionType);
+        this.writerFactory = writerFactoryCreator.apply(ManifestFileMeta.schema(partitionType));
+    }
+
+    public void write(List<ManifestFileMeta> metas, Path path) throws IOException {
+        Preconditions.checkArgument(
```
Comment (JingsongLi) [inline flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestListWriter.java]: We can have a `FileUtils.writeListToFile`?

Comment (tsreaper) [reply flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/manifest/ManifestListWriter.java]: We can't, at least currently. Most of the writers have some special logic (for example collecting the stats, calculating the number of added files) so it is hard to extract a utility method.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**